src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2574,15 +2641,18 @@ protected virtual void ProcessMethod (MethodDefinition method, in DependencyInfo
 
 			if (method.HasOverrides) {
 				foreach (MethodReference ov in method.Overrides) {
-					MarkMethod (ov, new DependencyInfo (DependencyKind.MethodImplOverride, method), method);
+					MarkMethod (ov, new DependencyInfo (DependencyKind.MethodImplOverride, method), new MessageOrigin (method));
 					MarkExplicitInterfaceImplementation (method, ov);
 				}
 			}
 
 			MarkMethodSpecialCustomAttributes (method);
 
-			if (method.IsVirtual)
+			if (method.IsVirtual) {
 				_virtual_methods.Add (method);
+				if (reason.Kind == DependencyKind.AccessedViaReflection)
+					ProcessRequiresUnreferencedCode (method, new MessageOrigin (reason.Source as IMemberDefinition), reason.Kind);
```

Comment:
This feels wrong - I understand it's because MarkEntireType doesn't call MarkMethod - maybe it should...
Can you also check if there are other places which side-step MarkMethod and end up here instead? Those could all be potential issues as well.